Map providers step2

tilaktilak opened the pull request:
Following this issue : #7774

[Rebased on the top of https://github.com/mavlink/qgroundcontrol/pull/7818 ]
[Diff between step 1 and step 2](https://github.com/Aeronavics/qgroundcontrol/compare/Aeronavics:MapProviders...Aeronavics:MapProviders_step2)

This is the backend to integrate custom DEM to QGC.
It contains the integration of GDAL, the ability to find tiff files in a folder and try to use them as an elevation data provider. If the current coordinates requested are not in the tiff files, fallback to Airmap Provider. 

The UI part needs to be done to connect this backend.
( at the moment a "Import DEM" button allow the user to select a folder, and search in that folder to find .tiff files : 
![image](https://user-images.githubusercontent.com/6662416/66530502-84df6280-eb64-11e9-9cae-351caae87776.png)
)

This was tested and works on Linux / Android (apk: 38Mb) / Windows

### Now with a significantly small amount of code/time we can add :
#### Elevation related
- Open a lot of filetypes that contains elevation data, see : https://gdal.org/drivers/raster/index.html
- For instance zipped hgt ( from SRTM mirror can be open by a "GDALOpen" function call with /vsi )
- Connect to a map provider that serves elevation tiles ( WMS / TMS / ... )

#### Maps related 
- Easily add/remove at runtime base map providers ( Custom tile server with a WMS / TMS / XYZ API )
- Think about having several map layers ( My orthophoto that overlay Google Satellite for instance).

@dogmaphobic 
@edwinhayes

dogmaphobic commented:
> The UI part needs to be done to connect this backend.

It's on my TODO list to rename "Offline Maps" to simply "Maps" and move the map settings (Provider and Type) to this (consolidated) map settings. Adding a button to define the path would go there. This should be fine for desktop. How do you plan on handling mobile devices?

tilaktilak commented:
That is great ! :-)
I've spend some time to cross-compile gdal on Android, I added a gdal folder under qgroundcontrol/libs with precompiled libs + scripts that I used to generate those libs. 
I was concerned about the size but my apk is 38Mb at the moment with libgdal + libproj embedded in it (gdal is compiled with a minimal amount of options though).
I also added dll for windows that comes from a GDAL release.

cosmarc commented on libs/gdal/android/build_gdal.sh:
```diff
@@ -0,0 +1,125 @@
+ --with-libtiff=internal \
+ --with-geotiff=internal \
+ --with-libjson-c=internal \
+ --with-proj \
```
Isn't there a need for the path to libproj here? 
`
 ./configure --help | grep proj
  --with-proj=ARG Compile with PROJ.x (ARG=no or path)
`

tilaktilak commented:
I tried to modify yml to compile with GDAL, I got :
 `‘Jenkinsfile’ has been modified in an untrusted revision`

Can someone help me on that ? Thanks !
@DonLakeFlyer

cosmarc commented:
@tilaktilak I would compile GDAL on windows with exactly the same modules enabled as on the other platforms (such that they provide similar functionality), so almost all of them disabled since that extra functionality will not be used and those many binaries will not be needed in the repo.

DonLakeFlyer commented:
@antiheavy You should follow along the work that @tilaktilak  is doing.

tilaktilak commented:
@cosmarc have you done any progress on your side ?
Also, GDAL need to be compiled for Android as it's a cross-compiled version, not released by GDAL. I kept the build "recipe" for posterity, but it's integrated the same way as other external libs in QGC. 
I think taking GDAL release as much as we can is a clean way to do and it's easy to maintain.
We don't want to build GDAL every time we build QGC as it's a fairly long process.

@DonLakeFlyer looks like I don't have the right to edit the pipeline config,

DonLakeFlyer commented:
> looks like I don't have the right to edit the pipeline config,

Not sure what you mean?

DonLakeFlyer commented:
> looks like I don't have the right to edit the pipeline config,

Wait probably Jenkins stuff? @dagar Can we turn off jenkins? Unless this truly becomes the default build it just creates busy work for everyone keeping it up to date.

tilaktilak commented:
From what I see Jenkins fallback on a old version of the Jenkinsfile because my branch is an untrusted source (which make totally sense otherwise everyone can use mavlink's jenkins runners to do what they want).
I may have a syntax error on Travis though.

DonLakeFlyer commented:
> I may have a syntax error on Travis though.

Yeah. If travis doesn't even show up as a check then that is usually what that means. They (travis) used to have a yml validator on their website somewhere. I think it's gone now though.

DonLakeFlyer commented:
Make sure you don't have tabs in .travis.yml just spaces. That hits me all the time.

tilaktilak commented:
So everything build but MacOS.

I need some help on this as I've no access to a MacOS machine.

I've tried to install gdal as a build dependency for Travis : 
- brew install take too long (actually build gdal) and the pipeline timeout

Might have the beginning of another solution here : 
https://trac.osgeo.org/gdal/wiki/DownloadingGdalBinaries#MacOSX

and here (See developer note) : https://www.kyngchaos.com/software/frameworks/ 

Looks like we can download a dmg that contain gdal and use it to build ? 
Also, it may be possible to create a "bottle" with brew and store it somewhere ?

Some hints here as well : https://trac.osgeo.org/gdal/wiki/BuildingOnMac

@DonLakeFlyer or @dogmaphobic  by any chance, can you help me on that ?  
Thanks

tilaktilak commented:
@DonLakeFlyer I manage to build this for Mac (Travis works now).

dogmaphobic commented:
What's the status on this?

tilaktilak commented:
Some conflict from this weekend merge, let me fix this. Other than that it's ready to merge.

tilaktilak commented:
@dogmaphobic Pipeline passed.

cosmarc commented:
@tilaktilak Can you please split this pull request in multiple ones but smaller?
I would split them something like this:
### These ones doesn't need so much review, just to check if there are some useless files (I already saw some, I think on Windows, GDAL is not correctly compiled). They mainly contain the binaries for GDAL and PROJ.
- binaries Windows
- binaries Mac
- binaries Android
- binaries Linux

### QGC code changes that would need more review
- changes of TerrainQuery, TerrainTile, QtLocation plugin, etc. (based on the binaries ones)

Otherwise it's super hard to review it and get it merged.
@dogmaphobic I would not merge this humongous PR.

tilaktilak commented:
So this is already a split PR from the previous MapProvider_step1 branch where you were involved in the PR.

> They mainly contain the binaries for GDAL and PROJ.
> 
>     binaries Windows
As discussed before with you, GDAL for windows is exactly the release from GDAL itself, it's not compiled.
>     binaries Mac
There is no binary for Mac
>     binaries Android
>     binaries Linux
There is no binary for Linux

It will look like a "humongous" PR anyway as the dependency is added in qgroundcontrol/libs like all dependencies for QGroundControl. 
I can't spend any more time on this PR.

cosmarc commented:
@tilaktilak : On Windows, you have here more binaries than needed. Here's how to compile GDAL for Windows:
## Compiling [PROJ](https://github.com/OSGeo/PROJ) 5.2.0 (needed by GDAL)

### **Common steps for all platforms:**

```bash
git clone https://github.com/OSGeo/PROJ.git
cd PROJ
git checkout 5.2.0
```
### **Windows** (Visual Studio 2015 32bit)

* Open `VS2015 X86 Native Tools Command Prompt`.
* Open `nmake.opt` and change `INSTDIR` variable to where the output libs will be copied in the end.
* **The paths should be absolute. Relative paths will mess up the submake files!**

```bash
C:\path\to\PROJ> nmake /f makefile.vc
C:\path\to\PROJ> nmake /f makefile.vc install-all
```
## Compiling [GDAL](https://github.com/OSGeo/gdal) 2.4.2

### **Common steps for all platforms:**

```bash
git clone https://github.com/OSGeo/gdal.git
cd gdal/gdal
git checkout v2.4.2
```
### **Windows** (Visual Studio 2015 32bit)

* Open `VS2015 X86 Native Tools Command Prompt`.
* Open `nmake.opt` and change `GDAL_HOME` variable to where the output libs will be copied in the end.
* **The paths should be absolute. Relative paths will mess up the submake files!**

```bash
set PROJ4_LIB=C:\path\to\PROJ\output\lib\proj.lib
C:\path\to\GDAL> nmake /f makefile.vc
C:\path\to\GDAL> nmake /f makefile.vc devinstall
```

tilaktilak commented:
About Windows, we have already been through that :
**Compile manually a single time a library is not a sane manner to add a dependency to QGC** as you can't know how it was compiled, with what dependencies, what environment... it's not reliable nor maintainable to do it this way. 
We need to use GDAL releases out of the box as much as we can as it's easy, reliable, future proof and it's the only way to have a tested version of the library. 
It's especially true here as Windows is the main platform for QGC users and we don't have any containerized build environment for that (only for Linux and Android). 
This project already uses pre-compiled libraries for all the complex dependencies for the same reasons: starting from Qt framework itself to GStreamer, OpenSSL, zlib, qtSerial, ...

Sorry but I don't have time to work on that anymore, we can decide to close this PR but
IMHO that stable enough to keep it and make it evolve afterward.

It's also a major feature for a "mapping-oriented" ground control software like QGC and that opens a lot of doors to : 
- Change the default elevation provider, use several providers. GDAL can already import zipped SRTM/AW3D30/... from a remote server using .vrt files in 1 line!
- Load Custom DEM data to ensure precise terrain following
- Ease the usage of any geospatial data of any kind ( to overlay DEM on a map and warn for terrain avoidance, load GCP files and show them on the map or use them to help plan the mission, ... )

This work of having generic interfaces for Map Provider and Elevation Provider makes us able to :
- Add Map/Elevation Provider easily, and at runtime to use with a custom TileServer (WMS/TMS/ ... )
- Overlay dataset on the map 
- Build easily a consistent UI to handle map sources/layers/...

tilaktilak commented:
Up on that @cosmarc @dogmaphobic, Go / No Go ?

cosmarc commented:
@tilaktilak I'll not merge it like it is now, but I worked on it in the past weeks and I'll try to find some time close to the end of the month to push some more  changes on top of this PR.

klck commented:
How far have things gone so far? Can someone explain whether theres still ongoing development on this topic?
I think this would be a phenomenal feature (especially importing my own .tiff elevation models) for relativley low altitude fligths while terrain following mode is activated.

Kind regards

jleppert commented:
Can someone explain how to use this feature? I built it and select my folder with my DEM tiff files, but the terrain visualization doesn't show and it doesn't appear to work. We want to merge this into our internal fork of QGC as we need this feature.

DonLakeFlyer commented:
Closing as abandoned

elijah-blake-bnc commented:
I'm on a team that's interested in the custom elevation functionality this PR was intended to provide. We'd like to incorporate it into a fork of QGC, or even finish the work and make a new PR to the QGC repo, if possible. Would one of the devs mind sharing why this was abandoned so far into development?

I would appreciate any background or information on the topic. Thank you in advance and sorry for the necro.

Vaakefjell commented:
Following. I live above 60°N, so Airmap elevation data doesn't cover my area. The ability to import it manually would be great, as I'm interested in mapping cliffs and geological features.

ronsand commented:
I really hope someone can pick up the thread and continue the excellent work started here on Map Providers, as being able to use different map providers and custom elevation data would really benefit all of that live North of 60°N
